unix_integration/src/idprovider/interface.rs

```diff
@@ -86,8 +91,52 @@ pub enum AuthCacheAction {
     PasswordHashUpdate { cred: String },
 }
 
+pub struct KeyStore<'a> {
+    dbtxn: &'a DbTxn<'a>,
+}
+
+impl<'a> KeyStore<'a> {
+    pub(crate) fn new(dbtxn: &'a DbTxn<'a>) -> Self {
+        KeyStore { dbtxn }
+    }
+
+    pub fn get_tagged_hsm_key<K: DeserializeOwned>(
+        &mut self,
+        tag: &str,
+    ) -> Result<Option<K>, IdpError> {
+        self.dbtxn
+            .get_tagged_hsm_key(tag)
+            .map_err(|_err| IdpError::KeyStore)
+    }
+
+    pub fn insert_tagged_hsm_key<K: Serialize>(
+        &mut self,
+        tag: &str,
+        key: &K,
+    ) -> Result<(), IdpError> {
+        self.dbtxn
+            .insert_tagged_hsm_key(tag, key)
+            .map_err(|_err| IdpError::KeyStore)
+    }
+
+    pub fn delete_tagged_hsm_key(&mut self, tag: &str) -> Result<(), IdpError> {
+        self.dbtxn
+            .delete_tagged_hsm_key(tag)
+            .map_err(|_err| IdpError::KeyStore)
+    }
+}
+
 #[async_trait]
 pub trait IdProvider {
+    async fn configure_hsm_keys(
+        &self,
+        _keystore: &mut KeyStore,
```

Comment:
Passing the keystore reference here is what causes the error.

Reply:
It's also possible you can wrap this in your own lock on your side, but you still may need to dance around lifetimes here.